### ISSUE TITLE
Knowledge Bases are not parsed correctly when doing Entity Linking

### ISSUE THREAD

**Issue by thetaaaaa** [inception-project/inception#5070]
**Describe the bug**
I imported a self-built ontology with instances as the knowlege base to do Entity Linking. But, I can only saw concepts which have more than one level, the others are missing.

**To Reproduce**
1. Create a new project based on the built-in Layer "Entity Linking" and import an custom ontology with all import options setted as default. In my case, the ontology is  [Ancient Oratory Ontology](https://talos-ai4ssh.uoc.gr/wp-content/uploads/2019/04/oratory_v.1.0-1.zip)
2. Check "Knowledge Base", and there are only concepts with more than one levels are visible, others with just one level are missing.
Picture below shows the difference between how the ontology are displayed in Inception and Protege.
![image](https://github.com/user-attachments/assets/7357d5ff-50ea-4557-9007-45a7fb0253c3)
![image](https://github.com/user-attachments/assets/34cbb1ce-907f-4655-a686-2788d947bf87)

**Expected behavior**
All concepts should be visible in Inception, with the same hierarchy as it in Protege.

**Screenshots**

**Please complete the following information:**
 - Version and build ID: 33.7 (2024-09-18 21:04:21, build c156ca28)
 - OS: Windows
 - Browser: Edge

**Additional context**

**Comment by reckart** [comment]
INCEpTION requires that concepts are explicitly defined to appear visible. However, `ontolex#Form`, `skos:Concept`, and `skos:ConceptScheme` are not defined in the ontology you link to - they are only used. Looks like Protegé then infers their existence and auto-generates an incomplete definition for them. Protegé has quite a few more capabilities than INCEpTION including the ability to do inference which INCEpTION does not offer.

Which entries from your ontology do you want to link against?

**Comment by thetaaaaa** [comment]
I want to link to the instances under the ontolex#Form and skos:Concept, but they are invisible, so I cannot choose them. What areinecessary elements that required to make the concept visible in Inception ? In other words, what do you mean "INCEpTION requires that concepts are explicitly defined to appear visible?"

Here is my test on my ontology: I add a property "rdfs:isDefinedBy:xxx" to the the ontolex#Form, but after that, “ontolex#Form” is still invisible in Inception. Below are pictures of the same modified ontology in Protege and Inception:
![image](https://github.com/user-attachments/assets/7bb25375-8363-418b-aaa0-55e482ead1a4)
![image](https://github.com/user-attachments/assets/6fd4d1d2-d307-4afb-a6e3-b06c3cdc2285)

**Comment by reckart** [comment]
I have imported the `oratory_v.1.0.rdf` into Protegé and then saved it again as RDF/XML so it includes the concepts that are auto-generated by Protegé. 

I set up two knowledge bases. 

The first I call `Oratory Lemon`, import the augmented ontology and I configure it like this:

<img width="1040" alt="Screenshot 2024-09-26 at 19 17 43" src="https://github.com/user-attachments/assets/17961a22-01fd-412e-bef4-0983a4358d8b">
<img width="1042" alt="Screenshot 2024-09-26 at 19 18 09" src="https://github.com/user-attachments/assets/f477fbaa-53ca-48ea-b7bf-9e2282ace1a0">
<img width="1035" alt="Screenshot 2024-09-26 at 19 18 18" src="https://github.com/user-attachments/assets/3e452a72-87b7-4471-92a0-507e0a8812c0">

This gives me this content:

<img width="1029" alt="Screenshot 2024-09-26 at 19 18 44" src="https://github.com/user-attachments/assets/b4ed3fd7-52ac-4578-997d-1fe79182a9ca">

The first I call `Oratory SKOS`, import the augmented ontology and I configure it like this:
<img width="1042" alt="Screenshot 2024-09-26 at 19 19 29" src="https://github.com/user-attachments/assets/c6bdac8d-fc23-45e8-aec6-bbedc6ca171e">

<img width="1035" alt="Screenshot 2024-09-26 at 19 19 39" src="https://github.com/user-attachments/assets/1408a3ae-4359-409c-9431-6e813de6251b">

Which then gives me this content:

<img width="1060" alt="Screenshot 2024-09-26 at 19 20 21" src="https://github.com/user-attachments/assets/a0a0feec-fc98-4f83-9dee-4c3e85543ac9">

Note that in the current v33 version, the "additional language" cannot be configured in the UI. If you need it, you can configure at by adding the following line to your `settings.properties` file:

```
knowledge-base.default-fallback-languages[0]=grc
```

If you want to use two different annotation features to each link to only one of the "views" of the data, you may configure two concept features and select for each one of the knowledge bases.

Does that work for you?

**Comment by thetaaaaa** [comment]
Thank you for your reply and effort. I followed your steps and get these results:
(1)Your second practice works the same on my device.
(2) Your first practice don't work on my device. I resave the "oratory_v.1.0.rdf" through Protege, and renamed it as "lemon". Then I imported "lemon" in to Inception, and try to set the Root Concepts with the URI of the concept "ontolex#Form", and Inception prompt me with the error below, so it seems that Inception still didn't find the concept "ontolex#Form" on my device.
Error shown by Inception:
![image](https://github.com/user-attachments/assets/203c3438-bbb8-4b0e-9cd0-32042add4239)

I checked the resaved ontology "lemon.rdf", it do has the concept"#Form" inside. Below is the screenshot of the resaved ontology:
![image](https://github.com/user-attachments/assets/41a00b58-ae4e-47da-99f4-5a50232f1a80)

The version of my Protege is 5.6.4.

**Comment by reckart** [comment]
For me, that worked. I have attached the RDF file I have exported from Protege here.

[oratory.rdf.zip](https://github.com/user-attachments/files/17155488/oratory.rdf.zip)

**Comment by reckart** [comment]
If you scroll up, what size information do you see. It should be like this:

<img width="1039" alt="Screenshot 2024-09-26 at 22 08 54" src="https://github.com/user-attachments/assets/2006b64b-1bc9-40a1-96f8-a76995ee9cf5">

**Comment by reckart** [comment]
Note, to import a file after creating a KB, you need to select the file in the import contents field and then press "Save" on the bottom. Only after that, setting the root will work.

**Comment by thetaaaaa** [comment]
> If you scroll up, what size information do you see. It should be like this:
> 
> <img alt="Screenshot 2024-09-26 at 22 08 54" width="1039" src="https://private-user-images.githubusercontent.com/1410238/371300589-2006b64b-1bc9-40a1-96f8-a76995ee9cf5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjczODQ0OTQsIm5iZiI6MTcyNzM4NDE5NCwicGF0aCI6Ii8xNDEwMjM4LzM3MTMwMDU4OS0yMDA2YjY0Yi0xYmM5LTQwYTEtOTZmOC1hNzY5OTVlZTljZjUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjZUMjA1NjM0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDA4MDcxNjFhYjhkZmJiNTA1NzlmMzJkOTdiMjVhZWNkYTU4MTA3OGM4MTNlM2ZjOTc3NzcwMDljOTE3NTBlMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.SW6iIDLAtdXqh3bR94UTZO1lLZNRimM3SFf7FQoAybA">

I imported the ontology which your attached during the Creating Knowlege Base Step, and pressed "Save". The result is like below, we are not the same:
![image](https://github.com/user-attachments/assets/f6ad5878-9843-4e4f-9126-affba7dc5a81)

**Comment by thetaaaaa** [comment]
I tried many times on creating Knowledge Base, but it still not work as expected. I record a video of my operations. I wish you can check it to see if I do something wrong.<iframe src="//player.bilibili.com/player.html?isOutside=true&aid=113206127234923&bvid=BV1drxHeyEFF&cid=26021857686&p=1" scrolling="no" border="0" frameborder="no" framespacing="0" allowfullscreen="true"></iframe>

**Comment by reckart** [comment]
Thanks for the video. In fact, in my instructions I forgot an essential step... I am very sorry for that.

I though that passing the ontology once through Protegé was all that I had done... however, I did one additional step.

I downloaded the lemon ontology from here `http://www.w3.org/ns/lemon/ontolex` and saved it as `ontolex.rdf`.

I then imported that file into the knowledge base as well using the **Import contents** field in the KB settings and saving.

That brings the number of statements in the KB up from `4764` to `5348 ` and also finally provides the missing `http://www.w3.org/ns/lemon/ontolex#Form` concept.

Actually you don't even need to pass the oratory file through Protegé in this case. You could also simply select both the files `oratory_v.1.0.rdf` and `ontolex.rdf` (using the CTRL or SHIFT button while clicking in the file selection dialog) during the creation of the knowledge base and then just set the root concept to `http://www.w3.org/ns/lemon/ontolex#Form` and the label concept to `http://www.w3.org/ns/lemon/ontolex#writtenRep` - that should already be sufficient.

**Comment by thetaaaaa** [comment]
It works now. But the concept"#Form" and the others cannot automaticlly be shown in Inception at the same time. Do [you mean](https://github.com/inception-project/inception/issues/5070#issuecomment-2377523801) # I need to create two knowlege bases at the same time ?

I figure out another way,which is manully add all the first level concepts to "Root Concepts",like the picture below, is this also a  correct operation ?
![image](https://github.com/user-attachments/assets/b6559e28-9fa8-454e-b3cd-62927092eafa)
After that, I will get this result:
![image](https://github.com/user-attachments/assets/6e83efb6-f13c-4503-8574-6200616641df)

**Comment by reckart** [comment]
If you can configure the a single KB that gives you access to all of the concepts you are interested in, it is fine. 

I suggested to set up two KBs because the SKOS hierarchy typically uses different properties to build the tree (e.g. `skos#broader`)  than e.g. an OWL/RDF class hierarchy. 

Mind that the root concepts settings just gives you a nicer way of browsing the KB. It does not really affect the linking auto-complete on the annotation page. If you want to limit that, you can set a **scope** on the concept feature.

Glad it works and I guess now you have a bunch of options to try out to see what fits your personal preference most.

**Comment by thetaaaaa** [comment]
Thank you. I got new problems......
I create a new project with the Basic Annotation template. I uploaded my CoNLL2002 file as the prelabeled corpus. I want to annotate both the named entities their relationships with concepts and properties in my ontology.
(1) It seems built-in "Named Entity" Layer doesn't allow users to user more than one KB to labeling text. As you can see in the picture below, I cannot create another feature at the same built-in "Named Entity" Layer,  is this a bug ? If so, creating two KBs may not be a good option.
![image](https://github.com/user-attachments/assets/7c3fd1fa-0600-486d-969a-d81e458879b7)
(2) I create a custom Layer called "Property", in order to annotate the relationship between two built-in "Named Entity" with the property in my ontology, my setting is shown below, at the mean time, I do import all the related ontology, such as `ontolex.rdf`:
![image](https://github.com/user-attachments/assets/9403fef4-b18a-463d-8593-edf5d1bc9b90)
But when I try to annotate the property between two named entities, the properties in my ontology are invisible. As is shown in the pictures below, I cannot see any properties of my properties:
![image](https://github.com/user-attachments/assets/d60d1486-9a45-407e-87ac-3d130e8a8eee)
![image](https://github.com/user-attachments/assets/06215937-0a79-4bf3-a741-3f51e9674275)
I checked the project Knowledge Base, and I found that the displayed properties, are incomplete and totally different from the propertis compared with it in Protege(I don't even know where the displayed properties come from......):
![image](https://github.com/user-attachments/assets/89cddd88-3fa3-444e-ac6e-e8427cbe36bf)
So my question is how can I annotate the relationship between two built-in "Named Entity" with the property in my ontology? 
Thank you for your time and instructions.

**Comment by reckart** [comment]
You cannot add features to the built-in **Named Entity** layer. The features of built-in layers can currently not be changed. This is not a bug. 

However, if you leave the KB selection on the feature on **All knowledge bases**, then you can select concepts from both of your configured KBs.

In order to annotate relations between entities with concepts from your knowledge base, you would need to create a custom relation layer. On that custom relation layer, you can create arbitrary features. However, you cannot export that in CoNLL layer on. If you want to export, you will have to use e.g. UIMA CAS JSON.

Does that answer your question?

**Comment by thetaaaaa** [comment]
Thank you for you instruction. But I still remains some problems when I do manully annotation.
1. The `Properties` of my own ontologies are not visible and cannot be selected, only `Classes` and `Instances` are visible and can be selected. I created a [vedio](【Bug Report of KB in INCEPTION】 https://www.bilibili.com/video/BV1KJxYe8EC8/?share_source=copy_web&vd_source=c269ff663afb842acff6599950467062) for your reference.
2. There are always inconsistency between the displayed ontologyies(the ontologies I uesed are [here](https://talos-ai4ssh.uoc.gr/resources/open-datasets/) for your reference) in Protege and INCEPTION, and I can't even find a pattern for their differences in order to modify my KBs. Is it possible to enhance the function of KB in INCETION?
3. When I leave the KB selection on the feature on All knowledge bases, any of them cannot be selected when I do manully annotation. You can also check it in the same [vedio](【Bug Report of KB in INCEPTION】 https://www.bilibili.com/video/BV1KJxYe8EC8/?share_source=copy_web&vd_source=c269ff663afb842acff6599950467062) above.

**Comment by reckart** [comment]
Thanks for the pointers to the ontologies. I don't know when I may find time to look into these though.

Thanks also for the videos. However, I find them a bit hard to follow without explanation. INCEpTION does not aim to be a copy of Protegé. INCEpTION wants to be able to use its IRI mappings to access a configurable concept and property hierarchy from a knowledge base. Depending on how the KB is built that may or may not be possible without inference. INCEpTION does not do inference though. There are also certain kinds of KB designs that are not supported by INCEpTION, e.g. those modelling the concept hierarchy using the `skos:narrower` property. 

The KB support in INCEpTION is improved over time step by step. At the moment, I do not know what exactly would need to be done for the ontologies you linked. It might be it is just a matter of properly configuring the IRI mapping. It might also be that they use a structure that is currently not supported and that code changes are necessary. 

It is easier for me to process focussed issue reports or questions related to a specific knowledge base.  The more you package into an issue, the more time it takes me to look into it, the less likely I am to be able to do anything about it in the near future.

**Comment by thetaaaaa** [comment]
Thank you for your reply and time and explaination! My objective is to use INCEpTION to create instances of a given Ontology by annotating a given raw text, that is the reason why I need all the property and concept in an ontology to be correct prased and displayed. I will keep exploring INCEpTION to see if I can solve the problem.
